domain/domain.go

```diff
@@ -409,10 +409,10 @@ type EtcdBackend interface {
 }
 
 // NewDomain creates a new domain. Should not create multiple domains for the same store.
-func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duration, factory pools.Factory, sysFactory func(*Domain) (pools.Resource, error)) (d *Domain, err error) {
+func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duration, factory pools.Factory, sysFactory func(*Domain) (pools.Resource, error)) (*Domain, error) {
```

Comment:
I think this code is a bit of chaos, which lead to the source of bugs.
Would it be better if we split `NewDomain` into two parts: create object, and then initialize?
`NewDomain()` will return the object, with no error.
`domain.Init()` will initialize all kind of things, if it return error, we **must** close the domain object.